### DISCUSSION TITLE
Adding Qt support to datoviz

### DISCUSSION THREAD

**Discussion by chittti** [datoviz/datoviz#45] [General]
This discussion is to track the addition of Qt support, at least with the C API to datoviz. 

I have modified datoviz, and I can currently render to a surface. In this effort I found out there are two possible approaches that can be taken:
1.  Make datoviz return a sub-classed QWindow or QWidget. This would mean additional dependencies on datoviz (which can be optional), and a bit of convoluted re-structuring since datoviz is primarily written in C. This does have a precedent, looking at imgui supported added in gui.cpp
2. Modify datoviz's API to work with user supplied VkSurfaceKHR and force the user to manage more of the event loop. This might come in the form of the user supplying a DvzWindow*, not just the surface. The benefit this would have is no dependencies added, and datoviz being ready to work with more backends since we're technically not just allowing user to embed datoviz in Qt apps but any situation where they are given access to a VkSurfaceKHR. 
 
@rossant, do you have any opinions provided your past experience? My effort successfully rendering in a Qt application was with (2). I did try to make (1) work, but there might be a little but more work getting a non-NULL surface before the surface is exposed to the user. It should be possible in theory, however.

**Comment by chittti** [comment]
I came across the other discussion where datoviz-distributed is mentioned, and remembered it will supersede the current repo. So I decided to take a look to see if the canvas creation API is similar so the work can be ported over. To my surprise, the new canvas API does take in a surface by default. 

Should I focus my effort into the new datoviz-distributed repo, considering that changes I make here might be obsolete (or even redundant if I go with (2))? What is the confidence level in the current architecture of datoviz-distributed being similar to what it would be at release?

**Comment by rossant** [comment]
Thank you for your efforts on this front. I would absolutely love to support Qt, in some way or another.

If you have to go into the internals of datoviz, then it might be wiser to work on the datoviz-distributed repo, even if it's still a work in progress. The low-level code is very different. For example, as you noticed, the new canvas API does take a surface by default.

In the (still totally undocumented..) new repo, there's a strong separation between the low-level, Vulkan-dependent code, and the higher-level scientific plotting API. So far, only the former has been more ore less done (it's ~90% there). The latter is still at the spec stage, but I hope I can make quick progress on it in the coming months.

Between the two, there's an asynchronous, thread-safe protocol where one can send visualization "requests" to the renderer, which processes them in the event loop (it's all happening in the same OS process). The extent of what you can do with the requests is relatively limited, which is a good thing as it reduces the communication surface between the two. You can essentially create new windows, create/resize/delete GPU buffers or textures, create builtin graphics pipelines (markers, paths, text, etc.), upload and download data to/from the GPU, and prepare rendering commands. With this small set of requests, you cover a large part of scientific plotting use-cases. Obviously, the richness and complexity of the architecture lies in the builtin library of supported graphics primitives. You can try to get an idea of how it works [by looking at this test perhaps](https://github.com/datoviz/datoviz-distributed/blob/main/tests/test_presenter.c#L149).

So far, the renderer was built around glfw, but I tried to isolate and minimize the dependencies as much as possible. I think what you want to do is a Qt renderer. The event loop will be handled by Qt and not by datoviz/glfw. You're going to have a Qt app that runs the event loop, waits for new requests, and processes them via `renderer.h`. You probably want to reuse as much of the renderer as possible. The existing window creation logic should probably be completely bypassed to use Qt Vulkan widgets instead.

This architecture is likely to remain relatively stable. I expect to implement only minor improvements while I make progress on the higher-level scene interface, which will be solely based on `request.h` and, that's the whole point, will not use any of the low-level internals stuff directly.

**Comment by chittti** [comment]
Yes, I don't think there is any way to add Qt support to the current datoviz repo without going into the internals, so I'm going to switch to the datoviz-distributed repo. Are any of the current datoviz examples functional on the new repo? I am most interested in getting [this example](https://datoviz.org/examples/signals/) working.

**Comment by rossant** [comment]
No, not yet. The closest example may be [this one](https://github.com/datoviz/datoviz-distributed/blob/main/tests/test_presenter.c#L661). You can probably use a basic graphics with LINE_STRIP primitive to reproduce this example. You'll have to bake the data to handle the transitions between consecutive polylines (see the corresponds graphics' baking function in the original datoviz repo). And for pan/zoom or any kind of interactivity/animation, you'll have to do everything manually by sending data upload requests to the correct buffers... All of this logic will be implemented as part of the higher-level scene interface.

**Comment by rossant** [comment]
This is now tracked in #42. In v0.3, there is experimental (and undocumented) support for limited PyQt integration using offscreen rendering and transferring the GPU image to the CPU as a Qt image widget every frame. Performance is better than expected, though there is room for optimization. Full Qt integration with a Qt Vulkan surface is planned for v0.4 (late 2025 or early 2026).
